Deprecate non-prefixed letters in warning specifications

Octachron opened the pull request:
This PR adds a new alert, `ocaml_deprecated_cli`, when using non-prefixed letter in a warning specification.
For instance, calling ocamlopt with
```bash
$ ocamlopt -w fragile-math test.ml
```
raises the following alert:
>```
>File "_none_", line 1:
>Alert ocaml_deprecated_cli: Setting a warning with a single lowercase or uppercase letter is deprecated.
>Prefix each letter with either '+' or '-': -a-e-f-g-h-i-l-r-t.
>```

This PR aims to make the coexistence of letter and named warning more harmonious. In particular, catching spelling mistake in warning name is already an useful feature. In the longer term, removing non-prefixed letter warning would make it possible to decouple the parsing of warnings from the list of known warnings at version `v`.

Using a specific deprecation alert seemed a good option, since this a deprecation of the cli interface which is quite distinct from a deprecation in the standard library.

Looking at the change in the testsuite, most of them correspond to switching `-w a` or `-w A` to `-w -a` and `-w +A` respectively.
It might make sense to make an exception for this specific case, but regularity has its charm.

cc @gasche , @damiendoligez

Octachron committed:
stop using single letter in warning settings

- Replace -w A with -w +A in various makefiles and tests.
- Replace -w a with -w -a in tests.
- Replace -w Ae with -w +A -w -fragile-matching

Octachron committed:
Fix caml_tex warning setting

Octachron committed:
compiler interface: deprecate single letter warning

For instance, -w "Ae" now raises an ocaml_deprecated_cli alert.

Setting warning with prefixed letters (-w "+a-e") is clearer and
coexists better with named warnings (-w +fragile-match).

In particular, this new alert should warn on spelling mistakes like
-w +fragile-math

gasche commented:
Thanks for the work! Ideally I would like to iterate on the message, from the perspective of "is it clear enough to someone who intended to write a warning name mnemonic?".

First suggestion:

> Alert ocaml_deprecated_cli: Setting a warning with a single lowercase or uppercase letter like 'f' is deprecated.
> You could use the equivalent form with '+' or '-', here for example -a-e-f-g-h-i-l-r-t.

I wonder if the letters in the example could be given in order of occurrence, rather than in alphabetic order. (So you would accumulate in a list instead of this `seen` array.)

Maybe we could also add a hint:

> Hint: if you want to enable or disable a warning by its name, the form is -<name> or +<name>.

I still am not sure how to get across the idea that the message is parsed in a different way than expected. Maybe just having in the deprecation message something like:

> The current warning specification uses sign-less single-letter warning specificiers ('f' corresponds to -'f', and 'F' to '+f").
> These forms are deprecated, you should use the equivalent signed form, in this case
>   -f-r-a-g-i-l-m-a-t-h.
> Hint: enabling/disabling a warning by its mnemonic name is also possible with -<name> or +<name>.

Octachron commented:
Keeping the user order is indeed better.
In term of global structure, I would rather keep the structure deprecation warning/proposed fix/hint.

Maybe:

> Alert ocaml_deprecated_cli: Setting a warning with single lowercase or uppercase letters, like 'f' or 'F', is deprecated.
> Use the equivalent signed form: -f-r-a-g-i-l-e-a-t-h
> Hint: Enabling/disabling a warning by its mnemonic name requires  a `+` or `-` prefix.
> Hint: Did you make a spelling mistake when using a mnemonic name?.

Giving an example of the deprecated specifier on the first line seems indeed clearer.
I agree with the idea of highlighting the equivalence between the two forms. I am less fond of personalizing the message with `you should`. 

Concerning the hint part, I think it is important to mention both the fact that mnemonic names require a prefix, and that the above message could have fired due to a spelling mistake. Eventually, I could add more logic to only emit one of those two hints when it is relevant. But it is probably better to do that once we have a stable syntax for specifying multiple named warnings.

gasche commented:
Your proposal is good. The second hint should be shown only above a certain number of sign-less specifiers (I would suggest 5).

gasche commented:
Note: in the proposed "equivalent form", I think it would be nicer to also include the non-obsolete flags, to give a sentence that really is equivalent to the full specification.

Note: once we all agree to allow spaces to separate specifiers, we will add spaces in-between for clarity.

Note: another approach would be to give locations within the specifier, with proper highlighting of the problematic part, and to have distinct warnings for non-disjoint deprecated segments of the full specifier. But this is overkill, harder to implement, and results in longer messages.

Octachron committed:
warning deprecation: full normalization of parsed tokens

The fix part of the deprecated message now provides a normalized
version of the parsed tokens.

Octachron committed:
Changes

Octachron commented:
I have changed the format, and switched to the fully normalized form in the proposed fix:
```bash
./ocamlopt.opt -I stdlib/ -w -fra@10..20-30+match
```
emits now:
>```
>Alert ocaml_deprecated_cli: Setting a warning with single lowercase or uppercase letters, like 'r' or 'R',
>is deprecated.
>Use the equivalent signed form: -f-r-a@10..20-30+m-a-t-c-h.
>Hint: Enabling or disabling a warning by its mnemonic name requires a + or - prefix.
>Hint: Did you make a spelling mistake when using a mnemonic name?
>```
Note that it seemed cleaner at this point to split the parsing and the interpretation in the warning DSL.

damiendoligez commented on utils/warnings.ml:
```diff
@@ -523,60 +566,89 @@ let parse_opt error active errflag s =
-let () = parse_options false defaults_w;;
-let () = parse_options true defaults_warn_error;;
+let () = ignore @@ parse_options false defaults_w;;
+let () =ignore @@ parse_options true defaults_warn_error;;
```
```suggestion
let () = ignore @@ parse_options true defaults_warn_error;;
```

damiendoligez commented on stdlib/Compflags:
```diff
@@ -27,10 +27,10 @@ case $1 in
-  stdlib__Buffer.cm[io]) echo ' -w A';;
-  camlinternalFormat.cm[io]) echo ' -w Ae';;
+  stdlib__Buffer.cm[io]) echo ' -w +A';;
+  camlinternalFormat.cm[io]) echo ' -w +A -w -fragile-match';;
```
Why not the straightforward translation to `+A-e`, if only as an example of quickly fixing existing code?

Octachron replied:
Essentially because I did not remember what was the `e` option.

damiendoligez reviewed (approved):
I have carefully reviewed all 318 changed files and I believe the changes are correct.

I'm still worried about the disruption to opam. Could you (@Octachron) do a run of opamcheck to get an idea of the amount of breakage?

If there is too much disruption, we could also consider adding a special case for `A` and `a`, which would probably cover a large part of the uses (opamcheck again?).

damiendoligez commented:
>Note: another approach would be to give locations within the specifier, with proper highlighting of the problematic part, and to have distinct warnings for non-disjoint deprecated segments of the full specifier. But this is overkill, harder to implement, and results in longer messages.

Overkill indeed !

Octachron commented:
I will run an opamcheck run over the next week(s).

gasche commented:
Non-update: we are still interested in the results of an opamcheck run.

dra27 commented:
I rebased the PR on to 4.12.0 and @kit-ty-kate has done an opam-health-check differential run. With a minor patch to `ppx_js_style` (https://github.com/kit-ty-kate/ppx_js_style/commit/dc84f5c7f11c8a331a902e50a838b6505298114d) and a similarly minor patch to utop (https://github.com/kit-ty-kate/utop/commit/41610c431bd2486252fdb7575e20dd3b4a587b96) we get [these results](http://check.ocamllabs.io:8081/?comp=4.12&comp=4.12%2Bpr10207&available=4.12&show-diff-only=true&show-latest-only=true&maintainers=&logsearch=&logsearch_comp=4.12).

`ocamlbrowser` also uses the changed `Warnings.parse_options` (but nothing depends on it, which is why Kate didn't patch it). The other failures look to me to be either testsuite-related or un-related to this change.

The opam-health-check logs are grep'able - it suggests that packages triggering this alert are either old ones of ours (`num`, `ocamlbuild`, ...) or otherwise actively maintained (`findlib`, `ocamlnet`, `batteries`, etc.). So it looks as though this PR only breaks 3 packages and we have an easy way to detect the packages which are triggering the alert to allow a speedy "completion" of the deprecation.

gasche commented:
It's amusing that the two patches are not related to the deprecation itself, but to the fact that `Warnings.parse_options` changes its output type (now non-unit).

I guess it would be nice to ping the maintainers of the broken packages (frama-c, gsl, lwt-exit, ocamlbrowser. ounit, p4pp). I could participate if no one is actively taking care of it.

(I received a patch from @kit-ty-kate for ocamlbuild indeed, which was using non-prefixed letters in some place. The fact that there exists a simple, backwards-compatible fix makes these patches trivial to review and accept.)

My impression is that the results are reassuring: very few packages are affected, and it's entirely manageable to fix them in time for 4.13; so we could/should go ahead to proceed with merging the change. Do other people agree with this assessment?

dra27 commented:
> My impression is that the results are reassuring: very few packages are affected, and it's entirely manageable to fix them in time for 4.13; so we could/should go ahead to proceed with merging the change. Do other people agree with this assessment?

On the basis of @damiendoligez having done a thorough review, and the health-check confirming that the opam breakage should be manageable, I agree, too! It would also be horrendous to have to rebase this, I expect...

Octachron commented:
At some point I was wondering if we should keep the same type for `Warning.parse_options` and add an extended version (`Warning.parse_options_with_deprecated_alerts`?) that returns the deprecated alert option. It would make things simpler for users the API, but it would make the deprecation alert reach fewer users.

dra27 commented:
I don't think it's necessary to have a separate function - it's a very small number of users, most of whom should at least consider doing something with the alert! It's very easy to do nothing (via `ignore`) and it's relatively easy if you need a compat wrapper for older compiler-libs to always return `None` (and you'd still do that even if there were `Warnings.parse_options_with_deprecated_alert`).

gasche reviewed (approved):
I propose to go ahead and merge, then. @Octachron, could you rebase to fix the last conflict?

lpw25 commented:
Just got bitten by this PR. Not a fan. Could we please keep `-w a` working, and ideally things like `-w a-1-2-3` too?

Octachron commented:
It seems fine to me to not trigger the deprecation warning on `a` and `A` and this would cover both of your examples . Or would you prefer the deprecation warning to trigger on any uses outside of `a+x+y+...` or `A-x-y-...` ?

kit-ty-kate commented:
@lpw25 do you have alerts treated as fatal errors in release mode or is it just a matter of having encountered this in dev mode?

lpw25 commented:
> Or would you prefer the deprecation warning to trigger on any uses outside of a+x+y+... or A-x-y-... ?

I don't have a strong opinion.

> do you have alerts treated as fatal errors in release mode or is it just a matter of having encountered this in dev mode?

I guess in "dev mode". I generally write `-w a` fairly frequently and I'm going to get repeatedly annoyed by this warning one way or another.
